PeerDAS (#1)
hwwhww/consensus-specs: specs/peerdas/p2p-interface.md

- [Modifications in PeerDAS](#modifications-in-peerdas)
  - [Custom types](#custom-types)
  - [Preset](#preset)
  - [Containers](#containers)
    - [`DataColumnSidecar`](#datacolumnsidecar)
  - [Helpers](#helpers)
      - [`get_row`](#get_row)
      - [`get_column`](#get_column)
      - [`verify_column_sidecar`](#verify_column_sidecar)
  - [The gossip domain: gossipsub](#the-gossip-domain-gossipsub)
    - [Topics and messages](#topics-and-messages)
      - [Samples subnets](#samples-subnets)
        - [`data_column_{subnet_id}`](#data_column_subnet_id)
  - [The Req/Resp domain](#the-reqresp-domain)
    - [Messages](#messages)
      - [DataRowByRootAndIndex v1](#datarowbyrootandindex-v1)
      - [DataColumnByRootAndIndex v1](#datacolumnbyrootandindex-v1)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
<!-- /TOC -->

## Modifications in PeerDAS

### Custom types

We define the following Python custom types for type hinting and readability:

| Name | SSZ equivalent | Description |
| - | - | - |
| `ExtendedData` | `ByteList[MAX_BLOBS_PER_BLOCK * BYTES_PER_BLOB * 2]` | The full data with blobs and 1-D erasure coding extension |

> | `DataRow`   | `ByteList[BYTES_PER_BLOB * 2]` | The data of each row in PeerDAS |

Reviewer: would we serve the entire extended row via `req/resp`? or can we just serve the `BlobSidecar` as we do in the existing `deneb` spec?

Author: removed it (https://github.com/hwwhww/consensus-specs/pull/1/commits/b5d12207c4e3d4ede71b81c645d4761191356083)

> | `DataColumn`   | `ByteList[MAX_BLOBS_PER_BLOCK * BYTES_PER_BLOB * 2 // NUMBER_OF_COLUMNS]` | The data of each column in PeerDAS |

Reviewer: what about defining a `DataCell` and then defining `DataColumn` to be a list of `DataCell`?

Author: FYI the List-form `DataColumn` has 24 extra bytes for the serialized object. But agreed it would simplify the spec. 

It's probably better to define it in a guide of erasure code and networking specs can reference it?

edited: address in b9702133a8d5c1d894fdabc7cc18f39b9120674c

> | `LineIndex`   | `uint64` | The index of the rows or columns in `ExtendedData` matrix |

Reviewer: if we can continue to use `BlobIndex` to specify a particular `BlobSidecar` (or extended blob), then could we use something like `DataColumnIndex` here? is there some forward compatibility concern?

Author: The benefit of not using `BlobIndex` is to not bind `NUMBER_OF_ROWS == MAX_BLOBS_PER_BLOCK` invariant with `DataRowByRootAndIndex` req. But I can see your point that it's more symmetric to use a  `DataColumnIndex` in `DataColumnSidecar`. 🤔

Reviewer 2: I agree with @jacobkaufmann on this and there is no `NUMBER_OF_ROWS` anymore.

### Preset

| Name                                     | Value                             | Description                                                         |
|------------------------------------------|-----------------------------------|---------------------------------------------------------------------|
| `KZG_COMMITMENTS_MERKLE_PROOF_INDEX`   | `uint64(get_generalized_index(BeaconBlockBody, 'blob_kzg_commitments'))` (= 27) | <!-- predefined --> Merkle proof index for `blob_kzg_commitments` |

### Containers

#### `DataColumnSidecar`

```python
class DataColumnSidecar(Container):
    index: LineIndex  # Index of column in extended data
    column: DataColumn
    kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    kzg_proofs: List[KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    signed_block_header: SignedBeaconBlockHeader
    kzg_commitment_merkle_proof: Vector[Bytes32, KZG_COMMITMENT_INCLUSION_PROOF_DEPTH]
```

### Helpers

##### `get_row`

```python
def get_row(data: ExtendedData, index: LineIndex) -> DataRow:
    length = BYTES_PER_BLOB * 2

>     assert len(data) % (BYTES_PER_BLOB * 2) == 0
>     assert len(data) // (BYTES_PER_BLOB * 2) <= MAX_BLOBS_PER_BLOCK

Author: ```suggestion
    assert len(data) % length == 0
    assert len(data) // length <= MAX_BLOBS_PER_BLOCK
```

    return data[index * length:(index + 1) * length]
```

##### `get_column`

```python
def get_column(data: ExtendedData, index: LineIndex) -> DataColumn:

>     assert len(data) % NUMBER_OF_COLUMNS = 0

Author: ```suggestion
    assert len(data) % NUMBER_OF_COLUMNS == 0
```

    assert BYTES_PER_BLOB * 2 % NUMBER_OF_COLUMNS == 0

    row_count = len(data) // NUMBER_OF_COLUMNS
    column_width = BYTES_PER_BLOB * 2 // NUMBER_OF_COLUMNS
    column = []
    for row in range(row_count):
        start = row * NUMBER_OF_COLUMNS + column_index
        column.append(data[start:start + column_width])
    return column
```

##### `verify_column_sidecar`

```python
def verify_column_sidecar(sidecar: DataColumnSidecar) -> bool:
    column = sidecar.column

>     column_width =  MAX_BLOBS_PER_BLOCK * BYTES_PER_BLOB * 2

Author: ```suggestion
    column_width =  BYTES_PER_BLOB * 2 // NUMBER_OF_COLUMNS
```

    assert len(cells) == len(sidecar.kzg_commitments) == len(sidecar.kzg_proofs)
    # KZG batch verify the cells match the corresponding commitments and proofs
    assert verify_cells(cells, sidecar.index, sidecar.kzg_commitments, sidecar.kzg_proofs)
    # Verify if it's included in the beacon block
    return is_valid_merkle_branch(
        leaf=hash_tree_root(data_line_sidecar.kzg_commitments),
        branch=data_line_sidecar.kzg_commitments_merkle_proof,
        depth=floorlog2(KZG_COMMITMENTS_MERKLE_PROOF_INDEX),
        index=KZG_COMMITMENTS_MERKLE_PROOF_INDEX,
        root=data_line_sidecar.signed_block_header.message.body_root,
    )
```

TODO: define `verify_cells` helper.

### The gossip domain: gossipsub

Some gossip meshes are upgraded in the fork of Pe to support upgraded types.

#### Topics and messages

##### Samples subnets

###### `data_column_{subnet_id}`

This topic is used to propagate column sidecars, where each column maps to some `subnet_id`.

The *type* of the payload of this topic is `DataColumn`.

TODO: add verification rules. Verify with `verify_column_sidecar`.

### The Req/Resp domain

#### Messages

> ##### DataRowByRootAndIndex v1

Author: Since the number of rows ==  `MAX_BLOBS_PER_BLOCK`, we can change it to just `BlobByRootAndIndex` and return `Blob` + compute extension locally.

Author: Removed it!

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Modifications in PeerDAS](#modifications-in-peerdas)
  - [Custom types](#custom-types)
  - [Preset](#preset)
  - [Containers](#containers)
    - [`DataLine`](#dataline)
    - [`DataLineSidecar`](#datalinesidecar)
  - [Helpers](#helpers)
      - [`verify_data_line_sidecar_inclusion_proof`](#verify_data_line_sidecar_inclusion_proof)
  - [The gossip domain: gossipsub](#the-gossip-domain-gossipsub)
    - [Topics and messages](#topics-and-messages)
      - [Samples subnets](#samples-subnets)
        - [`data_line_row_{line_index}`](#data_line_row_line_index)
        - [`data_line_column_{line_index}`](#data_line_column_line_index)
  - [The Req/Resp domain](#the-reqresp-domain)
    - [Messages](#messages)
      - [GetCustodyStatus v1](#getcustodystatus-v1)
      - [DASQuery v1](#dasquery-v1)
      - [DataLineQuery v1](#datalinequery-v1)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->
<!-- /TOC -->

## Modifications in PeerDAS

### Custom types

We define the following Python custom types for type hinting and readability:

| Name | SSZ equivalent | Description |
| - | - | - |

> | `DataLine`   | `ByteList[MAX_BLOBS_PER_BLOCK * BYTES_PER_BLOB]` | The data of each row or column in PeerDAS |

Reviewer: I think that we will want to differentiate between rows and columns. while it would be nice to have a unifying type, it appears to me that we will have to treat rows and columns differently with a 1-dimensional extension.

however, since we already have [`BlobSidecar`](https://github.com/ethereum/consensus-specs/blob/102809371e4bc04e71c269b836eea67aad87958b/specs/deneb/p2p-interface.md#blobsidecar), we may only have to introduce one new type to represent a column.

that type could be named something like `BlobColumnSidecar` or `BlobMatrixColumnSidecar`. the naming is less important to me, but maybe the suggestions are helpful.

that column type would likely benefit from some custom SSZ type to represent a "cell" or "component" of the column (e.g. vector whose length is equal to column width) along with a custom SSZ type to represent the entire column (e.g. vector of column cells/components).

the cell/component type could be the sampling unit.

Reviewer 3: I'm not 100% sure on if I'd do a different data-type yet. although I lean in favor.

That said, both could use the full `kzg_commitments` transmission with root-verification instead of the merkle proof, but would be moderately (but not extremely wrt full size of the sidecar) wasteful as we get to higher blob counts

Author: Changed it to:
1. Fully reuse BlobSidecar for rows
2. Add `DataColumnSidecar` for columns

@jacobkaufmann I stubbed the name `DataColumnSidecar` because the custom type is `DataColumn`. Personally I think "DataColumn" makes more sense to distinguish it's not just blobs. But of course, we can discuss it later.

### Preset

| Name                                     | Value                             | Description                                                         |
|------------------------------------------|-----------------------------------|---------------------------------------------------------------------|
| `KZG_COMMITMENTS_INCLUSION_PROOF_INDEX`   | `uint64(get_generalized_index(BeaconBlockBody, 'blob_kzg_commitments'))` (= 27) | <!-- predefined --> Merkle proof index for `blob_kzg_commitments` |

### Containers

#### `DataLine`

```python

> class SlotDataLine(Container):

Reviewer 3: We don't need this anymore right? Both rows and columns need a sidecar (either same or separate type)

Reviewer 3: Or if we use it in the req/resp, we should probably also use it in the DataLineSideCar (assuming we don't break this into row vs column types)

Author: Removed it

    slot: Slot
    data: DataLine
```

#### `DataLineSidecar`

```python
class DataLineSidecar(Container):
    start: BlobIndex

>     blobs: List[Blob, MAX_BLOBS_PER_BLOCK]

Reviewer 4: If a `DataLine` is a row or column, then why are they represented by a list of `Blob`s?

>     blobs: List[Blob, MAX_BLOBS_PER_BLOCK]
>     kzg_commitments: List[KZGCommitment, MAX_BLOBS_PER_BLOCK]  # All KZGCommitment in BeaconBlock

Author: ```suggestion
    blobs: List[Blob, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    kzg_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]  # All KZGCommitment in BeaconBlock
```

    signed_block_header: SignedBeaconBlockHeader

>     kzg_commitments_inclusion_proof: Vector[Bytes32, floorlog2(KZG_COMMITMENTS_INCLUSION_PROOF_INDEX)]

Reviewer 3: We don't need the inclusion proof anymore. You just compute the root of `kzg_commitments` and make sure that equals the root in the block header as well as verifying the blob(s) against the commitment(s)

Reviewer 3: Unless we do two different types for rows and columns. Rows can keep this structure

Reviewer: would you still need an inclusion proof for the column since the `blob_kzg_commitments` are in the `BeaconBlockBody`, and the `BeaconBlockHeader` only contains the `body_root`?

Reviewer: "inclusion proof" might not be the right name but something to check consistency between root of `blob_kzg_commitments` given in the column sidecar and the `body_root` in the `BeaconBlockHeader`

Author: @djrtwo Same question as @jacobkaufmann. I assume the data will be broadcast at a similar time as `BeaconBlock` broadcast. No?

> re: inclusion proof"

I was thinking about using the same naming pattern as the Deneb sidecar. can change it to "kzg_commitments_merkle_proof".

Author: Discussed with @djrtwo yesterday and the current Merkle proof verification is required.

```

### Helpers

##### `verify_data_line_sidecar_inclusion_proof`

```python

> def verify_data_line_sidecar_inclusion_proof(data_line_sidecar: DataLineSidecar) -> bool:

Reviewer 3: If we use the unified type where all kzg_commitments are transmitted, then we won't do a merkle proof. We'll just do a merkle tree computation (caching so just once per block/all messages) and check against the beacon root.

If we do separate type for rows vs columns, then this will still be used for rows.

Author: since I reused `BlobSidecar`, I also changed it to only verify columns.

    for i, blob in enumerate(data_line_sidecar.blobs):
        assert blob_to_kzg_commitment(blob) == data_line_sidecar.kzg_commitments[start + i]
 
    return is_valid_merkle_branch(
        leaf=hash_tree_root(data_line_sidecar.kzg_commitments),
        branch=data_line_sidecar.kzg_commitments_inclusion_proof,
        depth=floorlog2(KZG_COMMITMENTS_INCLUSION_PROOF_INDEX),
        index=KZG_COMMITMENTS_INCLUSION_PROOF_INDEX,
        root=data_line_sidecar.signed_block_header.message.body_root,
    )
```

### The gossip domain: gossipsub

Some gossip meshes are upgraded in the fork of Pe to support upgraded types.

#### Topics and messages

##### Samples subnets

###### `data_line_row_{line_index}`

> This topic is used to propagate the row data line and s

Reviewer 3: ```suggestion
This topic is used to propagate the row data line, where `line_index` maps to the index of the given column.
```

The *type* of the payload of this topic is `DataLineSidecar`. It contains extra fields for verifying if the blob(s) of the row is included in the given `BeaconBlockHeader` with Merkle-proof.

TODO: add verification rules

###### `data_line_column_{line_index}`

This topic is used to propagate the column data line, where `line_index` maps the index of the given column.

> The *type* of the payload of this topic is `SlotDataLine`.

Reviewer 3: This should be sidecar, right?

TODO: add verification rules

### The Req/Resp domain

#### Messages

##### GetCustodyStatus v1

> **Protocol ID:** `/eth2/beacon_chain/req/get_custody_status/1/`

Reviewer 3: I'm not certain we should do this independent bit-field check/reseeding on initial spec.

clients can poll neighbors if they are missing data. This is a nice-to-have on reseeding missing data, but is unnecessary to know *who* to poll given the deterministic custody functions

Author: removed the whole `GetCustodyStatus`

The `<context-bytes>` field is calculated as `context = compute_fork_digest(fork_version, genesis_validators_root)`:

Request Content:
```
(
  slot: Slot
)
```

Response Content:
```
(
  Bitvector[NUMBER_OF_ROWS * NUMBER_OF_COLUMNS]
)
```

The response bitfield indicates the samples of the given slot that the peer has and can provide.

> ##### DASQuery v1

Reviewer 3: Maybe `DataSampleBySlotAndIndex`.... a bit verbose, but might be useful to make sure we can disambiguate from other requests (e.g. blocks by range vs by root, instead of blockquery)

Reviewer 3: Also, we might want to make this a multi-query -- plural and a list of query params like the blocks requests

**Protocol ID:** `/eth2/beacon_chain/req/das_query/1/`

The `<context-bytes>` field is calculated as `context = compute_fork_digest(fork_version, genesis_validators_root)`:

Request Content:
```

> (

Reviewer 3: We might need to tighten this to a `block_root` so disambiguate if forking at the head

Author: Right, that makes more sense. And we can remove `slot`.

  slot: Slot
  sample_index: uint64
)
```

> `sample_index` maps the the index of the chunk in whole data in the flattened format (one-dimension).

Reviewer 3: Probably should write a quick reference mapping function `get_sample_index(line_type, line_index, index)` to disambiguate how this is compressed to a single integer index.

Or at least say in text -- indices move along rows placed in increasing order. or something

Author: @dankrad pointed out that in 1D construction, sampling units will have to be whole columns. I removed the "sample_index" term here and only use `DataColumn` now.

Response Content:
```
(
  ByteList[MAX_BLOBS_PER_BLOCK * BYTES_PER_BLOB * 2 // (NUMBER_OF_ROWS * NUMBER_OF_COLUMNS)]
)
```

> ##### DataLineQuery v1

Reviewer 3: Our naming convention on p2p doesn't usually prefix "get" or suffix "query".
Generally is just `DATA_TYPE` (and maybe a `by___` suffix). So I'd lean toward just calling this `DataLine` or `DataLineByTypeAndIndex` or something.

Also, we might want to do this a `DataLines` (plural) so you can batch multiple queries like our `byRoot` analogues

Author: This is based on the previous DAS specs - I somehow like it. But I'm fine with keeping following our naming convention.
